### DISCUSSION TITLE
Relation creation

### DISCUSSION THREAD

**Discussion by Fractal-Tess** [pocketbase/pocketbase#1153] [Q&A] [answered]
Hey there,

First of all, thank you very much for the hard work on the project. I only picked it up recently, and I can confirm that this is the ultimate side project database.

But I feel like there aren't enough examples on the website. In particular I'm talking about relations.

It could be my fault, as I am used to working with nosql databases, but I would love to see more examples of how to work with relations.

My question is about creating an entry in a collection that has a relation.

 Imagine two collections:
- movies
- genres

Given a completely empty collections, how would one insert movies, linking them to genres if that particular genre exists, or creating the genre if it does not. Would you please guide me through your thought process using the js-sdk. 

Thank you

**Comment by uteminhduc** [comment, marked as answer]
Try this:
**movies**
- title
- genres (relation)

**genres**
- name


```js
const client = $pocketbase()
const genre_names = ['genre1', 'genre2']
let genreIds = []
for (genre_name of genre_names) {
    const genre = await client.collection('genres').getFirstListItem(`name='${genre_name}'`)
    if (!genre) {
        genre = await client.collection('genres').create({name: genre_name})
    }
    genreIds.push(genre.id)
}

await client.collection('movies').create({
    title: "New title",
    genres: genreIds
  })
```
ref: [https://github.com/pocketbase/js-sdk#collectionservice](url)

**Comment by ganigeorgiev** [comment]
@Fractal-Tess The relations (single or multiple) are associated with their ids, eg:

```js
await pb.collection('example').create({
    singleRel: 'REL_RECORD_ID',
    multiRel: ['REL_RECORD_ID1', 'REL_RECORD_ID2', 'REL_RECORD_ID3'],
})
```
_You can find similar examples in the "API Preview" panel for your collection in the Admin UI._

The only thing that you have to be aware is that for multi-relation fields you always have to pass the final array of ids. For example, in the above if you decide to remove `'REL_RECORD_ID3'`, the update action should look something like this:

```js
await pb.collection('example').update('EXAMPLE_RECORD_ID', {
    multiRel: ['REL_RECORD_ID1', 'REL_RECORD_ID2'],
})
```
_In future version you'll be able to add/remove a single value from the multi-relation fields without explicitly setting the other rel values._

---

The @uteminhduc example is a good one, since it shows how to use a multi-relation field and how to deal with "upsert"-like scenarios client-side.

Another more traditional SQL-y approach would be to have a junction collection, for example you could have 3 collections - `movies`, `genres` and `movieGenreRels`, where the last one will store only the relations between the first 2. But this approach maybe is a little verbose for this simple use case.

---

To get a list with the movies and their genre records in a single request you can use the `?expand` query parameter (without `expand` you'll have access only to the genre ids):

```js
const result = await pb.collection("movies").getList(1, 30, { 'expand': 'genres' })
```
_More info and examples on `expand` you could find in https://pocketbase.io/docs/expanding-relations/._

**Reply by Fractal-Tess** [reply]
Thank you for the quick response, it clears a lot of what was fussy to me.

So if I'm getting this right, I need to loop overt the genres of a movie, check if it exists, create the genre if it doesn't and collect their ids in the process, only when I'm done with that, can I finally insert the movie with an array of genre ids.

That is what I originally had in mind as well, but it seemed kind of way too much repeater work. For example if I have 20 movies with 10 genres each, I would have to make 200 select requests to see if each genre exists, up to 200 create requests if none of the genres exists, and 20 more create requests to insert the movies.

To reduce this I'm thinking of keeping a in memory cache of the generes that already exists, to prevent querying for repeated ones, and perhaps to avoid many requests I could use the getList to get multiple.

Would you be able to confirm if this is the idiomatic approach one should use when dealing with such relations?

**Reply by ganigeorgiev** [reply]
@Fractal-Tess It depends whether you want to have a predefined list with genres records (eg. comedy, action, etc.) or you want also to allow custom user defined genres (aka. similar to creating new tags). uteminhduc's example shows the second option - it creates a new user defined genre (eg. "comedy2") if a genre with the provided name doesn't exist already.

**In both cases, the end goal is to have the relation ids so that you can pass them to the `movies` create/update request. How these ids will be fetched it's up to the developer and their use case.**

If the genres are finite and predefined, then Yes - you can fetch all genres upfront and extract the ids of those who matches your criteria.

But if the genres collection can grow indefinitely (eg. in the case of public user defined genres), then it may not be feasible to fetch all genres at once, because they could be thousands or more, and in this case the individual requests approach may be better (eventually you can also "batch" the requests by combining them in a single filter expression using the `&&` or `||` operators, eg. `name='genre1' || name='genre2' || ...`).

**Reply by Fractal-Tess** [reply]
Thank you!
